subsys/net/lib/app/net_app.c

```diff
@@ -509,11 +509,21 @@ struct net_context *select_client_ctx(struct net_app_ctx *ctx,
 		}
 	} else {
 		if (!dst) {
+			if (ctx->default_ctx->ctx &&
```

Comment:
> It might happen in TCP client, that the TCP connection is terminated in which case net_context is freed. Check this and mark corresponding net_context inside net_app to NULL. This way there will be no issue to access already freed net_context.

This is very dangerous. What if it's freed and already reused?

Such condition should not happen, it's a bug and we should be fix it (which should be as simple as incrementing a refcount for it, or not so simple).

I'm writing in particular because I see some weird issues in MicroPython which could be explained by this issue to.

Reply:
> What if it's freed and already reused?

That is why we mark the context as NULL here so that we do not reuse it in net-app.

I am not sure if adding refcount here would help as if TCP connection is closed, then we would need to inform the users of the net_context that it is closed. Probably using other flags in net_context we could figure this out in users of net_context.

Reply:
> That is why we mark the context as NULL here so that we do not reuse it in net-app.

What if it's reused by non-net-app?

> I am not sure if adding refcount here would help as if TCP connection is closed, then we would need to inform the users of the net_context that it is closed.

Well, (at the very least) we just need to make sure that the closed context is properly marked so (so any further operation on it will return appropriate error), but it's not freed until user closes ("puts") its own reference to it.